diff --git a/Makefile b/Makefile
index 75c7973..bae4143 100644
--- a/Makefile
+++ b/Makefile
@@ -18,6 +18,11 @@
 MAKEREC=$(MAKE)
 include Makefile.shared
 
+ifdef OCAML_LANDMARKS
+COMPFLAGS+=-I $(LANDMARKS_SRC) -ppx $(LANDMARKS_SRC)/ppx_landmarks
+LINKFLAGS+=-I $(LANDMARKS_SRC) landmarks.cma
+endif
+
 SHELL=/bin/sh
 MKDIR=mkdir -p
 
@@ -374,7 +379,7 @@ partialclean::
 ocamlnat: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
     otherlibs/dynlink/dynlink.cmxa compilerlibs/ocamlopttoplevel.cmxa \
     $(OPTTOPLEVELSTART:.cmo=.cmx)
-	$(CAMLOPT) $(LINKFLAGS) -linkall -o ocamlnat \
+	$(CAMLOPT) $(LINKFLAGS:.cma=.cmxa) -linkall -o ocamlnat \
 	    otherlibs/dynlink/dynlink.cmxa compilerlibs/ocamlcommon.cmxa \
 	    compilerlibs/ocamloptcomp.cmxa compilerlibs/ocamlopttoplevel.cmxa \
 	    $(OPTTOPLEVELSTART:.cmo=.cmx)
@@ -463,7 +468,7 @@ partialclean::
 
 ocamlc.opt: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamlbytecomp.cmxa \
             $(BYTESTART:.cmo=.cmx)
-	$(CAMLOPT) $(LINKFLAGS) -ccopt "$(BYTECCLINKOPTS)" -o ocamlc.opt \
+	$(CAMLOPT) $(LINKFLAGS:.cma=.cmxa) -ccopt "$(BYTECCLINKOPTS)" -o ocamlc.opt \
 	  compilerlibs/ocamlcommon.cmxa compilerlibs/ocamlbytecomp.cmxa \
 	  $(BYTESTART:.cmo=.cmx) -cclib "$(BYTECCLIBS)"
 
@@ -479,7 +484,7 @@ partialclean::
 
 ocamlopt.opt: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
               $(OPTSTART:.cmo=.cmx)
-	$(CAMLOPT) $(LINKFLAGS) -o ocamlopt.opt \
+	$(CAMLOPT) $(LINKFLAGS:.cma=.cmxa) -o ocamlopt.opt \
 	   compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
 	   $(OPTSTART:.cmo=.cmx)
 
diff --git a/debugger/Makefile b/debugger/Makefile
index d0ac256..3092624 100644
--- a/debugger/Makefile
+++ b/debugger/Makefile
@@ -15,3 +15,7 @@
 
 UNIXDIR=../otherlibs/$(UNIXLIB)
 include Makefile.shared
+ifdef OCAML_LANDMARKS
+COMPFLAGS+=-I $(LANDMARKS_SRC) -ppx $(LANDMARKS_SRC)/ppx_landmarks
+LINKFLAGS+=-I $(LANDMARKS_SRC) landmarks.cma
+endif
diff --git a/ocamldoc/Makefile b/ocamldoc/Makefile
index 2eba14c..5a49195 100644
--- a/ocamldoc/Makefile
+++ b/ocamldoc/Makefile
@@ -89,6 +89,11 @@ INCLUDES=$(INCLUDES_DEP) $(INCLUDES_NODEP)
 
 COMPFLAGS=$(INCLUDES) -warn-error A -safe-string
 LINKFLAGS=$(INCLUDES) -nostdlib
+ifdef OCAML_LANDMARKS
+COMPFLAGS+=-I $(LANDMARKS_SRC) -ppx $(LANDMARKS_SRC)/ppx_landmarks
+LINKFLAGS+=-I $(LANDMARKS_SRC) landmarks.cma
+endif
+
 
 CMOFILES= odoc_config.cmo \
 	odoc_messages.cmo\
@@ -188,19 +193,19 @@ debug:
 	$(MAKE) OCAMLPP=""
 
 $(OCAMLDOC): $(EXECMOFILES)
-	$(OCAMLC) -o $@ -linkall unix.cma str.cma dynlink.cma \
+	$(OCAMLC) -o $@ -linkall $(LINKFLAGS) unix.cma str.cma dynlink.cma \
 	          $(OCAMLSRCDIR)/compilerlibs/ocamlcommon.cma \
-	          $(LINKFLAGS) $(OCAMLCMOFILES) $(EXECMOFILES)
+	          $(OCAMLCMOFILES) $(EXECMOFILES)
 $(OCAMLDOC_OPT): $(EXECMXFILES)
-	$(OCAMLOPT) -o $@ -linkall unix.cmxa str.cmxa dynlink.cmxa \
+	$(OCAMLOPT) -o $@ -linkall $(LINKFLAGS:.cma=.cmxa) unix.cmxa str.cmxa dynlink.cmxa \
 	            $(OCAMLSRCDIR)/compilerlibs/ocamlcommon.cmxa \
-	            $(LINKFLAGS) $(OCAMLCMXFILES) $(EXECMXFILES)
+	            $(OCAMLCMXFILES) $(EXECMXFILES)
 
 $(OCAMLDOC_LIBCMA): $(LIBCMOFILES)
-	$(OCAMLC) -a -o $@ $(LINKFLAGS) $(OCAMLSRCDIR)/tools/depend.cmo \
+	$(OCAMLC) -a -o $@ $(LINKFLAGS:landmarks.cma=) $(OCAMLSRCDIR)/tools/depend.cmo \
 	          $(LIBCMOFILES)
 $(OCAMLDOC_LIBCMXA): $(LIBCMXFILES)
-	$(OCAMLOPT) -a -o $@ $(LINKFLAGS) $(OCAMLSRCDIR)/tools/depend.cmx \
+	$(OCAMLOPT) -a -o $@ $(LINKFLAGS:landmarks.cma=) $(OCAMLSRCDIR)/tools/depend.cmx \
 	            $(LIBCMXFILES)
 
 manpages: stdlib_man/Pervasives.3o
diff --git a/otherlibs/dynlink/Makefile b/otherlibs/dynlink/Makefile
index 8f57196..f29a79c 100644
--- a/otherlibs/dynlink/Makefile
+++ b/otherlibs/dynlink/Makefile
@@ -34,6 +34,10 @@ else
 OPTCOMPFLAGS=
 endif
 
+ifdef OCAML_LANDMARKS
+COMPFLAGS+=-I $(LANDMARKS_SRC) -ppx $(LANDMARKS_SRC)/ppx_landmarks
+LINKFLAGS+=-I $(LANDMARKS_SRC) landmarks.cma
+endif
 OBJS=dynlinkaux.cmo dynlink.cmo
 
 COMPILEROBJS=\
@@ -83,7 +87,7 @@ dynlink.cmx: dynlink.cmi natdynlink.ml
 	rm -f dynlink.mlopt
 
 extract_crc: dynlink.cma extract_crc.cmo
-	$(OCAMLC) -o extract_crc dynlink.cma extract_crc.cmo
+	$(OCAMLC) $(LINKFLAGS) -o extract_crc dynlink.cma extract_crc.cmo
 
 INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
 
diff --git a/tools/Makefile b/tools/Makefile
index 0e91277..0c04e15 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -15,6 +15,9 @@
 
 include Makefile.shared
 
+COMPFLAGS+=-I $(LANDMARKS_SRC) -ppx $(LANDMARKS_SRC)/ppx_landmarks
+LINKFLAGS+=-I $(LANDMARKS_SRC) landmarks.cma
+
 # To make custom toplevels
 
 ocamlmktop: ocamlmktop.tpl ../config/Makefile
diff --git a/tools/Makefile.shared b/tools/Makefile.shared
index 1a24391..f399264 100644
--- a/tools/Makefile.shared
+++ b/tools/Makefile.shared
@@ -52,7 +52,7 @@ ocamldep: depend.cmi $(CAMLDEP_OBJ)
 	         $(CAMLDEP_OBJ)
 
 ocamldep.opt: depend.cmi $(CAMLDEP_OBJ:.cmo=.cmx)
-	$(CAMLOPT) $(LINKFLAGS) -o ocamldep.opt $(CAMLDEP_IMPORTS:.cmo=.cmx) \
+	$(CAMLOPT) $(LINKFLAGS:.cma=.cmxa) -o ocamldep.opt $(CAMLDEP_IMPORTS:.cmo=.cmx) \
 	           $(CAMLDEP_OBJ:.cmo=.cmx)
 
 # ocamldep is precious: sometimes we are stuck in the middle of a
@@ -220,7 +220,7 @@ read_cmt: $(READ_CMT)
 	$(CAMLC) $(LINKFLAGS) -o read_cmt $(READ_CMT)
 
 read_cmt.opt: $(READ_CMT_OPT)
-	$(CAMLOPT) $(LINKFLAGS) -o read_cmt.opt $(READ_CMT_OPT)
+	$(CAMLOPT) $(LINKFLAGS:.cma=.cmxa) -o read_cmt.opt $(READ_CMT_OPT)
 
 clean::
 	rm -f read_cmt read_cmt.opt
@@ -284,7 +284,7 @@ OBJINFO=../compilerlibs/ocamlcommon.cma \
         objinfo.cmo
 
 objinfo: objinfo_helper$(EXE) $(OBJINFO)
-	$(CAMLC) -o objinfo $(OBJINFO)
+	$(CAMLC) $(LINKFLAGS) -o objinfo $(OBJINFO)
 
 install::
 	cp objinfo "$(INSTALL_BINDIR)/ocamlobjinfo$(EXE)"
